src/lib/plugins/sast/checks.ts

```
@@ -23,3 +28,18 @@ export async function getSastSettingsForOrg(org): Promise<SastSettings> {
 
   return (response as any).body;
 }
+
+export async function trackUsage(org): Promise<TrackUsageResponse> {
+  const response = await request({
+    method: 'POST',
+    headers: {
+      Authorization: `token ${snyk.api}`,
+    },
+    qs: assembleQueryString({ org }),
+    url: `${config.API}/track-sast-usage/cli`,
+    gzip: true,
+    json: true,
+  });
+
+  return (response as any).body;
```

Why do you need `as any` here? Can we make this type-safe? It looks like this response only has `userMessage` and `code` so it should not be difficult to parse.

Also it looks like you're not handling HTTP response codes (if the request fails) which will lead to an ambigious error. Maybe on non-ok status codes, there should be a fallback response with a user-friendly `userMessage`?

---

Wouldn't it be better to make this fail silently? It's only checking for the status code 429?

---

@JackuB I guess if the `body` is validated and falls back to a predefined body (whether that's an okay or a different failure, I'll leave that to SAST team to decide), it should handle any potential server issues. Something like:

```ts
try {
  return getTrackUsageResponse(response) // validate `code` and `userMessage` types
} catch (e) {
  debug(e)
  return okayTrackingUsageResponse
}
```

My main concern here is that `body` is not being validated so it can potentially be anything which defeats the point of using TypeScript.

---

There's (at least 😛 ) two problems here:
- we had a weird cyclic import for the api token which required! us to use require and thus we didn't have types -> fixed
- some of the types were incorrect (because not properly checked) -> fixed

What remains is because `request.ts` doesn't use generics we still need to assert the reponse type here, right?

---

Yes, but you can make it typesafe instead of using `any`. This might need a larger discussion (we use `any` a lot in places) so I won't block on this.